### PR TITLE
[search] Added Subtract() methods to BitVectors.

### PR DESCRIPTION
**Pull request by ygorshenin**
https://trello.com/c/1Wus49iN/37--

### CODE REVIEW

**Comment by mapsmetest** [comment]
Jenkins asks: can anyone have a look at this request?
If you want Jenkins to compile and test this, write "test this please" or JT_TP (without the underscore in the middle).

File: coding/compressed_bit_vector.hpp [hunk 1]
```diff
@@ -158,14 +164,14 @@ class CompressedBitVectorBuilder
       case CompressedBitVector::StorageStrategy::Dense:
       {
         vector<uint64_t> bitGroups;
-        ReadPrimitiveVectorFromSource(src, bitGroups);
+        rw::ReadVectorOfPOD(src, bitGroups);
```
Comment (mpimenov) [inline coding/compressed_bit_vector.hpp]: Тогда функцию ```ReadPrimitiveVectorFromSource``` вообще можно убрать.

File: coding/compressed_bit_vector.hpp [hunk 2]
```diff
@@ -1,15 +1,11 @@
-#include "std/vector.hpp"
-
-#include "base/assert.hpp"
-
+#include "coding/read_write_utils.hpp"
 #include "coding/reader.hpp"
 #include "coding/writer.hpp"
 
 #include "std/algorithm.hpp"
+#include "std/vector.hpp"
```
Comment (mpimenov) [inline coding/compressed_bit_vector.hpp]: sort

File: coding/compressed_bit_vector.cpp [hunk 3]
```diff
@@ -1,80 +1,175 @@
...
+      uint64_t const kBitsBegin = i * DenseCBV::kBlockSize;
+      uint64_t const kBitsEnd = (i + 1) * DenseCBV::kBlockSize;
+
+      uint64_t mask = 0;
+      for (; bit < b.End() && *bit < kBitsEnd; ++bit)
+      {
+        ASSERT_GREATER_OR_EQUAL(*bit, kBitsBegin, ());
+        mask |= static_cast<uint64_t>(1) << (*bit - kBitsBegin);
+      }
+
```
Comment (mpimenov) [inline coding/compressed_bit_vector.cpp]:
```
if (bit == b.End())
  break;
```
?

File: coding/compressed_bit_vector.cpp [hunk 4]
```diff
@@ -1,80 +1,175 @@
...
+  virtual unique_ptr<coding::CompressedBitVector> operator()(coding::SparseCBV const & a,
+                                                             coding::DenseCBV const & b) const = 0;
+  virtual unique_ptr<coding::CompressedBitVector> operator()(coding::SparseCBV const & a,
+                                                             coding::SparseCBV const & b) const = 0;
+};
 
-// The intersection of dense and sparse is always sparse.
-unique_ptr<coding::CompressedBitVector> IntersectImpl(coding::DenseCBV const & a,
-                                                      coding::SparseCBV const & b)
+struct IntersectOp : public BinaryOp
```
Comment (mpimenov) [inline coding/compressed_bit_vector.cpp]: Как насчёт статического полиморфизма?

File: coding/compressed_bit_vector.hpp (hunk 1 above)
Comment (ygorshenin) [inline coding/compressed_bit_vector.hpp]: Done

File: coding/compressed_bit_vector.hpp (hunk 2 above)
Comment (ygorshenin) [inline coding/compressed_bit_vector.hpp]: Done

File: coding/compressed_bit_vector.cpp (hunk 3 above)
Comment (ygorshenin) [inline coding/compressed_bit_vector.cpp]: Done

File: coding/compressed_bit_vector.cpp (hunk 4 above)
Comment (ygorshenin) [inline coding/compressed_bit_vector.cpp]: Done

**Comment by ygorshenin** [comment]
PTAL

**Comment by mpimenov** [comment]
LGTM

**Pull request merged**